### PR TITLE
Add JPA implementions of UserDao/ProfileDao

### PR DESCRIPTION
**Pull request by voievodin**
Along with JPA implementations i added 2 password encryptors _SHA-512_ and _PBKDF2_.
`JpaUserDao` and `JpaProfileDao` pass all the TCK prepared before. 

@skabashnyuk, @sleshchenko, @garagatyi, @akorneta  can you please review.
Please note that this will be merged into the `jpa-integration` branch which will be merged to the master when all the DAOs are migrated to JPA.

### CODE REVIEW

File: core/che-core-api-jdbc-vendor-h2/src/main/java/org/eclipse/che/api/core/jdbc/jpa/vendor/h2/H2ExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.eclipse.persistence.exceptions.ExceptionHandler;
+
+import static org.eclipse.che.api.core.jdbc.DBUtil.extractSqlErrorCode;
+
+/**
+ * @author Yevhenii Voevodin
+ */
+public class H2ExceptionHandler implements ExceptionHandler {
+
+    // TODO move it from here
```
Comment (sleshchenko) [inline core/che-core-api-jdbc-vendor-h2/src/main/java/org/eclipse/che/api/core/jdbc/jpa/vendor/h2/H2ExceptionHandler.java]: When do you plan remove it?

File: pom.xml [hunk 2]
```diff
@@ -42,11 +42,19 @@
     <properties>
         <che.lib.version>4.5.0-RC1-SNAPSHOT</che.lib.version>
         <che.version>4.5.0-RC1-SNAPSHOT</che.version>
+        <com.h2database.version>1.4.186</com.h2database.version>
```
Comment (sleshchenko) [inline pom.xml:45]: Do you plan move these dependencies to che-dependencies project?

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/model/impl/ProfileImpl.java [hunk 3]
```diff
@@ -21,19 +31,36 @@
...
+    @MapKeyColumn(name = "name")
+    @Column(name = "value")
+    @CollectionTable(joinColumns = @JoinColumn(name = "user_id"),
+                     uniqueConstraints = @UniqueConstraint(columnNames = {"user_id", "name"}))
     private Map<String, String> attributes;
 
-    public ProfileImpl(String id, String email) {
-        this.id = id;
+    // jpa mapping delegated to 'user' field
+    @Transient
```
Comment (benoitf) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/model/impl/ProfileImpl.java]: I don't get why we need the field if it's never stored

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/model/impl/UserImpl.java [hunk 4]
```diff
@@ -22,14 +28,29 @@
...
  */
+@Entity(name = "User")
 public class UserImpl implements User {
 
-    private String       id;
-    private String       email;
-    private String       name;
-    private String       password;
+    @Id
+    @Column(name = "id")
```
Comment (benoitf) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/model/impl/UserImpl.java]: why do we set the name of the columns and not let the default ?

**Comment by benoitf** [comment]
few remarks: 
- If there are password encryptors, why they're not part of a separated commits as it's not linked to JPA ?
- why the wheel has been re-invented with TCK modules ? while people are using Arquilian when testing JPA ? http://arquillian.org/guides/testing_java_persistence/
  - che-core-api-jdbc-vendor-h2
    the name is not obvious as it links the module to EclipseLink, while JPA allow to plug any persistence provider. so it should include that it's specific to eclipselink and is not "jdbc" neutral
- persistence.xml shouldn't be in wsmaster/che-core-api-user/src/main/resources/META-INF/
  it's part of the final assembly as it's where we select the persistence provider, and the database 
- Using container managed instead of JavaSE would have helped as here you've to managed transaction instead of only using annotations
  Using an autocloseable (or an interceptor like it's done in container managed world) should help to avoid to have tons of lines doing. Then you could pickup the transaction strategy (REQUIRED, REQUIRES_NEW, etc)

```

try {
            manager.getTransaction().begin();

// do something

           manager.getTransaction().commit();
        } finally {
            if (manager.getTransaction().isActive()) {
               manager.getTransaction().rollback();
            }
            manager.close();
       }
```

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/model/impl/ProfileImpl.java (hunk 3 above)
Comment (voievodin) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/model/impl/ProfileImpl.java]: Email is stored by `User` table while it is used in profile as well, there is no need to store email so the mapping is delegated to the User's field, on the other hand we need to create new profile instances from another profile instances or any data, in this way we need email field, which allows to avoid creation of embedded user object.

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/model/impl/UserImpl.java (hunk 4 above)
Comment (voievodin) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/model/impl/UserImpl.java]: Are you talking about `id` field or about the whole class mapping?
I am wondering that if we change field name the mappings will be also changed which requires
 migration as well. So it looks more strict for me, and i think it is necessary to name the columns explicitly. Your proposal?

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/model/impl/ProfileImpl.java (hunk 3 above)
Comment (benoitf) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/model/impl/ProfileImpl.java]: in which case do we need to create a ProfileImpl directly with a userId and a given email instead of getting email through the user object ?

If Profile has a JPA relationship with User object, there is no need to store/wrap email in any case as you can do getProfile().getUser().getEmail() as well

**Comment by voievodin** [comment]
> If there are password encryptors, why they're not part of a separated commits as it's not linked to JPA ?

np, i will extract ecnryptors code to the different commit.

> why the wheel has been re-invented with TCK modules ? while people are using Arquilian when testing JPA ? http://arquillian.org/guides/testing_java_persistence

Current TCK approach allows us use the same tests for different implementations Local/Ldap/Jpa.
Can Arquilian resolve this?

> che-core-api-jdbc-vendor-h2 the name is not obvious as it links the module to EclipseLink, while JPA allow to plug any persistence provider. so it should include that it's specific to eclipselink and is not "jdbc" neutral

_che-core-api-jdbc_ module contains common error codes and error codes mapper astraction used by daos, _che-core-api-jdbc-vendor-h2_ provides an implementation of error codes mapper specific for h2, while error codes are jdbc errors.
I will move code specific to eclipse link to the dedicated _jpa/eclipselink_ package are you okay with that? 

> persistence.xml shouldn't be in wsmaster/che-core-api-user/src/main/resources/META-INF/
> it's part of the final assembly as it's where we select the persistence provider, and the database

okay thank you, i am using it for test purposes only, so i will remove it from there.

> Using container managed instead of JavaSE would have helped as here you've to managed transaction instead of only using annotations
> Using an autocloseable (or an interceptor like it's done in container managed world) should help to avoid to have tons of lines doing. Then you could pickup the transaction strategy (REQUIRED, REQUIRES_NEW, etc)

We will use [this annotation](https://github.com/google/guice/wiki/Transactions), i will introduce it with next pull request, right now alot of things depend on this pull request that's why we decided to merge it asap which allows me and guys to work on jpa implementations separately.

File: pom.xml (hunk 2 above)
Comment (voievodin) [inline pom.xml:45]: yes

**Comment by benoitf** [comment]
> > If there are password encryptors, why they're not part of a separated commits as it's not linked to JPA ?
> 
> np, i will extract ecnryptors code to the different commit.

thx

> > why the wheel has been re-invented with TCK modules ? while people are using Arquilian when testing JPA ? http://arquillian.org/guides/testing_java_persistence
> 
> Current TCK approach allows us use the same tests for different implementations Local/Ldap/Jpa.
> Can Arquilian resolve this?

It allows to test with different JPA providers and different database as well
also it's a Container Agnostic framework and allow plugability.

> > che-core-api-jdbc-vendor-h2 the name is not obvious as it links the module to EclipseLink, while JPA allow to plug any persistence provider. so it should include that it's specific to eclipselink and is not "jdbc" neutral
> 
> _che-core-api-jdbc_ module contains common error codes and error codes mapper astraction used by >daos, _che-core-api-jdbc-vendor-h2_ provides an implementation of error codes mapper specific for h2, >while error codes are jdbc errors.
> I will move code specific to eclipse link to the dedicated _jpa/eclipselink_ package are you okay with >that? 

sure

> > persistence.xml shouldn't be in wsmaster/che-core-api-user/src/main/resources/META-INF/
> > it's part of the final assembly as it's where we select the persistence provider, and the database
> 
> okay thank you, i am using it for test purposes only, so i will remove it from there.

ok

> > Using container managed instead of JavaSE would have helped as here you've to managed >>transaction instead of only using annotations
> > Using an autocloseable (or an interceptor like it's done in container managed world) should help to >>avoid to have tons of lines doing. Then you could pickup the transaction strategy (REQUIRED, >>REQUIRES_NEW, etc)
> 
> We will use [this annotation](https://github.com/google/guice/wiki/Transactions), i will introduce it with >next pull request, right now alot of things depend on this pull request that's why we decided to merge it >asap which allows me and guys to work on jpa implementations separately.

Yes this Guice stuff or better, CDI, which offer more stuff around Java EE component will allow more possibilities around PersistenceContext

File: core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        requireNonNull(password, "Required non-null encrypted password");
+        final Matcher matcher = PWD_REGEX.matcher(encryptedPassword);
+        if (!matcher.matches()) {
+            return false;
+        }
+        // retrieve salt, password hash and iterations count from hash
+        final Integer iterations = tryParse(firstNonNull(matcher.group("iterations"), ""));
+        final String salt = matcher.group("saltHash");
+        final String pwdHash = matcher.group("pwdHash");
+        if (salt == null || iterations == null || iterations <= 0 || pwdHash == null) {
```
Comment (garagatyi) [inline core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java]: Encrypted password matches regexp and regexp require 1+ symbols in each group. It looks like NULL checks are not needed.

Comment (garagatyi) [inline core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java]: I mean salt and pwdHash

File: core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+ * Encrypts password using <a href="https://en.wikipedia.org/wiki/PBKDF2">Password-based-Key-Derivative-Function</a>
+ * with <b>SHA512</b> as pseudorandom function.
+ * See <a href="https://www.ietf.org/rfc/rfc2898.txt">rfc2898</a>.
+ *
+ * @author Yevhenii Voevodin
+ */
+public class PBKDF2PasswordEncryptor implements PasswordEncryptor {
+
+    private static final String  PWD_FMT   = "%s:%s:%d";
+    private static final Pattern PWD_REGEX = Pattern.compile("(?<pwdHash>\\w+):(?<saltHash>\\w+):(?<iterations>\\w+)");
```
Comment (garagatyi) [inline core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java]: Are you sure that iterations section may contain letters?

File: core/che-core-api-core/src/main/java/org/eclipse/che/security/SHA512PasswordEncryptor.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import java.security.SecureRandom;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * SHA-512 based encryptor {@code hash = sha512(password + salt) + salt}.
+ *
+ * @author Yevhenii Voevodin
+ */
+public class SHA512PasswordEncryptor implements PasswordEncryptor {
```
Comment (garagatyi) [inline core/che-core-api-core/src/main/java/org/eclipse/che/security/SHA512PasswordEncryptor.java:27]: Why do we need 2 implementations of password encryptor?

File: core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    @Override
+    public boolean test(String password, String encryptedPassword) {
+        requireNonNull(password, "Required non-null password");
+        requireNonNull(password, "Required non-null encrypted password");
+        final Matcher matcher = PWD_REGEX.matcher(encryptedPassword);
+        if (!matcher.matches()) {
+            return false;
+        }
+        // retrieve salt, password hash and iterations count from hash
+        final Integer iterations = tryParse(firstNonNull(matcher.group("iterations"), ""));
```
Comment (garagatyi) [inline core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java]: Since non-ASCII characters are not allowed here please check that also in encrypt method

File: core/che-core-api-core/src/test/java/org/eclipse/che/security/PasswordEncryptorsTest.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import static org.testng.Assert.assertTrue;
+
+/**
+ * @author Yevhenii Voevodin
+ */
+public class PasswordEncryptorsTest {
+
+    @Test(dataProvider = "encryptorsProvider")
+    public void testEncryption(PasswordEncryptor encryptor) throws Exception {
+        final String password = "password";
```
Comment (garagatyi) [inline core/che-core-api-core/src/test/java/org/eclipse/che/security/PasswordEncryptorsTest.java:26]: I think that this test doesn't cover several corner cases

File: core/che-core-api-jdbc/src/main/java/org/eclipse/che/api/core/jdbc/jpa/DetailedRollbackException.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.core.jdbc.jpa;
+
+import org.eclipse.che.api.core.jdbc.DBErrorCode;
+
+import javax.persistence.RollbackException;
+
+/**
+ * @author Yevhenii Voevodin
```
Comment (garagatyi) [inline core/che-core-api-jdbc/src/main/java/org/eclipse/che/api/core/jdbc/jpa/DetailedRollbackException.java:20]: Please add description of class into javadocs

File: core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java (hunk 8 above)
Comment (garagatyi) [inline core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java]: Skip this comment. My error.

File: core/che-core-api-core/src/main/java/org/eclipse/che/security/SHA512PasswordEncryptor.java (hunk 7 above)
Comment (voievodin) [inline core/che-core-api-core/src/main/java/org/eclipse/che/security/SHA512PasswordEncryptor.java:27]: Well, first i created _classic_ `SHA512PasswordEncryptor`  but then found a lot of articles recommending bcrypt/scrypt/PBKDF2 as safer mechanisms for password encryption, so i implemented another `PBKDF2PasswordEncryptor` with sha-512 as pseudorandom function. For `PBKDF2PasswordEncryptor` ecnryption time depends on the count of the iterations but `SHA512PasswordEncryptor` works faster, so i decided to not remove it as it is useful in tests.
Notice time:
image

**Comment by voievodin** [comment]
Pull request updated, @benoitf @skabashnyuk please check

File: core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java (hunk 8 above)
Comment (garagatyi) [inline core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java]: I think group can't be Null in that case

**Comment by benoitf** [comment]
ok

File: core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java (hunk 8 above)
Comment (voievodin) [inline core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java]: yes, thank you!

File: pom.xml (hunk 2 above)
Comment (skabashnyuk) [inline pom.xml:45]: I think it's good time to start creating CQs

Comment (voievodin) [inline pom.xml:45]: :C ok

**Comment by skabashnyuk** [comment]
ok

**Comment by garagatyi** [comment]
LGTM
